### DISCUSSION TITLE
Why does `annotate` not work with `input.http`?

### DISCUSSION THREAD

**Discussion by Moonbase59** [savonet/liquidsoap#3998] [Q&A]
For use in AzuraCast, we’d like to add some metadata to an `input.http`, like the internal playlist id.

When I try
```
playlist_remote = mksafe(buffer(buffer=10., input.http('annotate:playlist_id=217:https://example.com/radio.mp3')))

radio = playlist_remote
```

this produces an error
```
2024/06/30 08:25:37 [input.http:2] Decoding failed: Avutil.Error(Protocol not found)
```

**Is there an easy way to use annotate, so that we could use PHP to add some metadata to this?**

I know I could use something like
```
def metadata_remote(m) = [...m, ("playlist_id", "217")] end
playlist_remote = mksafe(buffer(buffer=10., input.http(metadata_filter=metadata_remote, 'https://example.com/radio.mp3')))

radio = playlist_remote
```
but that seems overdoing it a little, because we’d have to have one function per playlist, and add all that to the LS configuration.

**Comment by toots** [comment]
Internally, for liquidsoap, (request) protocols and `annotate:` (a request protocol) are for request resolution.

The `protocol` mention in the failed logs refers in fact to `ffmpeg`'s notion of [protocols](https://ffmpeg.org/ffmpeg-protocols.html), which is different than us.

Generally speaking, Liquidsoap requests are tracks that are acquired through rounds of resolutions. Each resolution is expected to last a finite time, with a timeout. The eventual result is a local file that can be played.

Live operators such as `input.http`, `input.ffmpeg` and etc. are designed to play a source as that is potentially infinite such as a icecast stream.

If you want to inject metadata, the generic methods, indendent of the nature of your source are `insert_metadata` and `metadata.map`. If you do like the `annotate` way of passing them, you can even parse a `annotate:` string using [string.annotate.parse](https://www.liquidsoap.info/doc-2.2.5/reference.html#string.annotate.parse)
